program/rust/src/time_machine_types.rs

```diff
@@ -3,56 +3,147 @@ use crate::c_oracle_header::{
+    pub running_valid_epoch_counter:   [u64; NUM_BUCKETS],
+}
+
+impl<const NUM_ENTRIES: usize> SmaTracker<NUM_ENTRIES> {
+    pub fn time_to_epoch(&self, time: i64) -> Result<usize, OracleError> {
+        try_convert::<i64, usize>(time / self.granularity)
+    }
+
+    pub fn initialize(&mut self, granularity: i64, threshold: u64) {
+        self.threshold = threshold;
+        self.granularity = granularity;
+        self.current_epoch_is_valid = true;
+    }
+
+    pub fn add_datapoint(&mut self, datapoint: &DataPoint) -> Result<(), OracleError> {
+        let epoch_0 = self.time_to_epoch(datapoint.last_two_timestamps.0)?;
+        let epoch_1 = self.time_to_epoch(datapoint.last_two_timestamps.1)?;
+
+        let datapoint_numerator = i128::from(datapoint.slot_gap) * i128::from(datapoint.price); //Can't overflow because u64::MAX * i64::MAX = i28::MAX
+        let datapoint_denominator = datapoint.slot_gap;
+
+        // Can't overflow, it's always smaller than the current solana slot
+        self.current_epoch_denominator += datapoint_denominator;
+        // Can't overflow, it's always smaller than u64::MAX * i64::MAX = i28::MAX
+        self.current_epoch_numerator += datapoint_numerator;
+        self.current_epoch_is_valid =
+            self.current_epoch_is_valid && datapoint_denominator <= self.threshold;
+
+        // This for loop is highly inefficient, but this is what the behavior should be
+        // It updates all the epochs that got skipped
+        // This will always yield an estimate of the average, even for invalid epochs
+        for i in epoch_0..epoch_1 {
+            self.conclude_epoch_and_initialize_next(i, datapoint_numerator, datapoint_denominator);
+            self.current_epoch_is_valid = datapoint_denominator <= self.threshold;
+        }
+        Ok(())
+    }
+
+    pub fn conclude_epoch_and_initialize_next(
+        &mut self,
+        epoch: usize,
+        datapoint_numerator: i128,
+        datapoint_denominator: u64,
+    ) {
+        let index = epoch.rem_euclid(NUM_ENTRIES);
+        let prev_index = (epoch + NUM_ENTRIES - 1).rem_euclid(NUM_ENTRIES);
+
+        // This buffer will be able to support u64::MAX epochs, no one will be alive when it
+        // overflows, that's because every epoch we add a number that's smaller than i64::MAX
+        self.running_sum_of_price_averages[index] = self.running_sum_of_price_averages[prev_index]
+            + self.current_epoch_numerator / i128::from(self.current_epoch_denominator);
+
+        // Likewise can support u64::MAX
+        // epochs
+        if self.current_epoch_is_valid {
+            self.running_valid_epoch_counter[index] =
+                self.running_valid_epoch_counter[prev_index] + 1;
+        } else {
+            self.running_valid_epoch_counter[index] = self.running_valid_epoch_counter[prev_index]
+        };
+
+        self.current_epoch_denominator = datapoint_denominator;
+        self.current_epoch_numerator = datapoint_numerator;
```

Comment:
i'm not sure i follow this update. I would have expected both of these to get set to zero. I think this is double (really multiple) counting updates that span epoch boundaries?

Reply:
Talked about in the call.